Add Conversational Search changes for 2.11 (#5195)
opensearch-project/documentation-website: _ml-commons-plugin/conversational-search.md

Currently, conversational search uses two systems to synthesize documents:

- [Conversation memory](#conversation-memory)

> - [Retrieval Augmented Generation (RAG) pipeline](#rag-pipeline)

Reviewer: Ha!  How did we miss this one?!  We need a spell checker!

Author: It happens from time to time!

## RAG pipeline

RAG is a technique that retrieves documents from an index, passes them through a seq2seq model, such as an LLM, and then generates more factual outputs.

> As of OpenSearch 2.11, the RAG technique has only been tested with the OpenAI and Bedrock Anthropic Claude models.

Author: ```suggestion
As of OpenSearch 2.11, the RAG technique has only been tested with OpenAI models and the Anthropic Claude model on Amazon Bedrock.
```

    "description": "This is a public model group"
  }
  ```

>   {% include copy-curl.html %}

Reviewer 2: On line 272, at and after the step "2. Create a new model group for the connected model. You'll use the `model_group_id` returned by the Register API to register the model:" the numbering is incorrect (because of tabs)

	"connector_id": "f5-iSYoBu0R6vVqGI3PA"
  }
  ``` 
  {% include copy-curl.html %}

4. With the model registered, use the `task_id` returned in the registration response to get the `model_id`. You'll use the `model_id` to deploy the model to OpenSearch:

>   ```curl

Reviewer 2: Throughout the page, please make sure that:
1. The requests are formatted correctly (some are squished to the left, others have really big tabs) 
2. Copy buttons are added to all requests
3. You use "json" for code blocks. "curl" is not a supported syntax highlighter in Jekyll.

        "tag": "openai_pipeline_demo",
        "description": "Demo pipeline Using OpenAI Connector",
        "model_id": "<model_id>",
        "context_field_list": ["text"],

>         "system_prompt": "You are a helpful assistance",

Author: ```suggestion
        "system_prompt": "You are a helpful assistant",
```

You can customize `context_field_list` in your RAG pipeline to send any fields that exist in your documents to the LLM.

### RAG parameter options

> Use the following options when setting up a RAG pipeline under the `retrieval_augmented_generation` argument:

Reviewer 2: ```suggestion
The following table lists the options the `retrieval_augmented_generation` parameter takes.
```

> Use the following options when setting up a RAG pipeline under the `retrieval_augmented_generation` argument:

Author: ```suggestion
Use the following options when setting up a RAG pipeline under the `retrieval_augmented_generation` argument.
```

Parameter | Required | Description
:--- | :--- | :---

> `tag` | Yes | A tag to help identify the pipeline.

Reviewer 2: Tag is optional in all other processors. Is it not optional here?

> `description` | Yes | A description of the pipeline.

Reviewer 2: Same. Description is optional in all other processors.

`model_id` | Yes | The ID of the model used in the pipeline.
`context_field_list` | Yes | The list of fields in document sources that the pipeline uses as context for the RAG. For more information, see [Context Field List](#context-field-list).

> `system_prompt` | No | The message sent to the LLM with a `system` role. This is the message the user sees when the LLM receives an interaction.

Reviewer 2: So if it's not required, what does the user see if system prompt is not provided?

Author: Whatever prompt was provided in the interaction.

Author: This is an override for the interactions set in the document or in Conversation API.

You can customize `context_field_list` in your RAG pipeline to send any fields that exist in your documents to the LLM.

### RAG parameter options

Use the following options when setting up a RAG pipeline under the `retrieval_augmented_generation` argument.

Parameter | Required | Description
:--- | :--- | :---

> `tag` | Yes | A tag to help identify the pipeline.

Author: ```suggestion
`tag` | No | A tag to help identify the pipeline.
```

`llm_question` | Yes | The question the LLM must answer. 
`llm_model` | No | Overrides the original model set in the connection in cases where you want to use a different model (for example, GPT 4 instead of GPT 3.5). This option is required if a default model is not set during pipeline creation.
`conversation_id` | No | Integrates conversation memory into your RAG pipeline by adding the 10 most recent conversations into the context of the search query to the LLM. 
`context_size` | No | The number of search results sent to the LLM. This is typically needed in order to meet the token size limit, which can vary model to model. Alternatively, you can use the `size` parameter in the Search API to control the amount of information sent to LLMs.

> `interaction_size` | No | The number of interactions sent to the LLMs. Similar to the number of search results, this affects the total number of token seen by the LLM. When set, the pipeline uses the default interaction size, `10`.

Reviewer 2: ```suggestion
`interaction_size` | No | The number of interactions sent to the LLMs. Similar to the number of search results, this affects the total number of tokens seen by the LLM. When not set, the pipeline uses the default interaction size, `10`.
```

> `timeout` | No | The number of seconds the pipeline waits for the remote model using a connector to respond. Default is `30`.

Reviewer 2: Is "using a connector" necessary here or can we remove and simplify?

Author: The pipeline will only work with a remote model connected through the model connector.

> `timeout` | No | The number of seconds the pipeline waits for the remote model using a connector to respond. Default is `30s`.

Reviewer: 30, not 30s.

> `timeout` | No | The number of seconds the pipeline waits for the remote model using a connector to respond. Default is `30s`.

Author: ```suggestion
`timeout` | No | The number of seconds the pipeline waits for the remote model using a connector to respond. Default is `30`.
```

`llm_question` | Yes | The question the LLM must answer. 
`llm_model` | No | Overrides the original model set in the connection in cases where you want to use a different model (for example, GPT 4 instead of GPT 3.5). This option is required if a default model is not set during pipeline creation.
`conversation_id` | No | Integrates conversation memory into your RAG pipeline by adding the 10 most recent conversations into the context of the search query to the LLM. 

> `context_size` | No | The number of search results sent to the LLM. This is typically needed in order to meet the token size limit, which can vary model to model. Alternatively, you can use the `size` parameter in the Search API to control the amount of information sent to LLMs.

Reviewer 3: ```suggestion
`context_size` | No | The number of search results sent to the LLM. This is typically needed in order to meet the token size limit, which can vary by model. Alternatively, you can use the `size` parameter in the Search API to control the amount of information sent to the LLM.
```

> `interaction_size` | No | The number of interactions sent to the LLMs. Similar to the number of search results, this affects the total number of tokens seen by the LLM. When not set, the pipeline uses the default interaction size, `10`.

Reviewer 3: ```suggestion
`interaction_size` | No | The number of interactions sent to the LLM. Similarly to the number of search results, this affects the total number of tokens seen by the LLM. When not set, the pipeline uses the default interaction size of `10`.
```

> `timeout` | No | The number of seconds the pipeline waits for the remote model using a connector to respond. Default is `30`.

Reviewer 3: ```suggestion
`timeout` | No | The number of seconds that the pipeline waits for the remote model using a connector to respond. Default is `30`.
```
